### PR TITLE
Add quote pattern bindings

### PR DESCRIPTION
**Pull request by nicolasstucki**
Introduces `Bind[T]` which can be used to match a check is an `Expr` is a reference to some
other binding defined in scope.

```scala
  case '{ val $x: Int = ($body: Int) } => // where x: Bind[Int]
  case '{ ($x: Int) => ($body: Int) } => // where x: Bind[Int]
  case '{ ($x: Int) => ($body: Int) + x } => // where $x is the name of the parameter
  case Bind(b) => // where b: Bind[Int]
```

- [x] Add `patternBindHole` annotation to names staring with `$` in quoted patterns
- [x] Handle `patternBindHole` in `Matcher`
- [x] Implement `Bind`
- [x] Distinguish names with back-ticks
- [x] Support names starting with `$` in lambda parameters

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Rebased

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -1973,6 +1974,22 @@ class Typer extends Namer
             val pat1 = if (patType eq patType1) pat else pat.withType(patType1)
             patBuf += pat1
           }
+        case ddef: ValOrDefDef =>
+          if (ddef.symbol.annotations.exists(_.symbol == defn.InternalQuoted_patternBindHoleAnnot)) {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: Simplify to `if (ddef.symbol.hasAnnotation(defn.InternalQuoted_patternBindHoleAnnot))`

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -1973,6 +1974,22 @@ class Typer extends Namer
...
+            val tpe = ddef.symbol.info match {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: Pick a more descriptive name for `tpe`. Maybe `bindingType`?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Renamed

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Simplified

**Pull request merged**